File: src/ServerB/UserServer.java
package serverB;

import utils.ORBSetup;
import src.AsistenciaModule.*;
import org.omg.CORBA.ORB;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.POAHelper;

public class UserServer {
    public static void main(String[] args) {
        try {
            // Inicializa el ORB
            ORB orb = ORBSetup.initializeORB(args);

            // Obtén una referencia al POA y actívala
            POA rootPOA = POAHelper.narrow(orb.resolve_initial_references("RootPOA"));
            rootPOA.the_POAManager().activate();

            // Crea una instancia del servicio y registra su implementación
            UserServiceImpl userService = new UserServiceImpl();

            // Obtén la referencia al objeto CORBA
            org.omg.CORBA.Object ref = rootPOA.servant_to_reference(userService);
            AsistenciaModule.UserService href = UserServiceHelper.narrow(ref);

            // Registra el servicio en el Naming Service
            org.omg.CORBA.Object namingContextObj = orb.resolve_initial_references("NameService");
            org.omg.CosNaming.NamingContextExt namingContext = org.omg.CosNaming.NamingContextExtHelper.narrow(namingContextObj);

            org.omg.CosNaming.NameComponent[] name = namingContext.to_name("UserService");
            namingContext.rebind(name, href);

            System.out.println("UserService listo y registrado...");

            // Mantén el servidor corriendo
            orb.run();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/server/user_server.py
import sys
from omniORB import CORBA
import CosNaming
from omniORB.CORBA import BOA
from omniORB import PortableServer

# Definición del servicio
class UserServiceImpl:
    def __init__(self):
        self.usuarios = {"Gandhi", "Arturo", "Juan", "Maria", "Pedro", "Jose", "Ana"}

    def validarUsuario(self, idUsuario):
        return idUsuario in self.usuarios


def main():
    try:
        # Inicializa el ORB
        orb = CORBA.ORB_init(sys.argv, CORBA.ORB_ID)
        poa = orb.resolve_initial_references("RootPOA")
        poa_manager = poa._get_the_POAManager()
        poa_manager.activate()

        # Crea el servicio
        servant = UserServiceImpl()

        # Registra el objeto en el POA
        obj_ref = poa.servant_to_reference(servant)

        # Obtiene el servicio de nombres
        naming_context = orb.resolve_initial_references("NameService")
        naming_context = naming_context._narrow(CosNaming.NamingContext)

        # Crea el nombre del servicio y lo registra
        name = "UsuarioService"
        name_component = [CosNaming.NameComponent(name, "")]
        naming_context.rebind(name_component, obj_ref)

        print(f"{name} registrado en el servicio de nombres.")
        print("Servidor listo y esperando conexiones...")

        # Ejecuta el ORB
        orb.run()
    except Exception as e:
        print(f"Error al configurar el servidor: {e}")


if __name__ == "__main__":
    main()
